File: tests/system/env_wrapper.cpp
/*
 * SPDX-FileCopyrightText: Copyright (c) 2023-2026 NVIDIA CORPORATION & AFFILIATES. All rights reserved.
 * SPDX-License-Identifier: Apache-2.0
 */

#include "env_wrapper.hpp"

#include <cstdlib>
#include <string>
#include <utility>
#include <vector>

#include "gxf/logger/gxf_logger.hpp"
#include "holoscan/logger/logger.hpp"

EnvVarWrapper::EnvVarWrapper(
    const std::vector<std::pair<std::string, std::string>>& env_var_settings)
    : env_var_settings_(env_var_settings) {
  // Save existing environment variables and apply new ones
  for (const auto& [env_var, value] : env_var_settings_) {
    const char* orig_value = getenv(env_var.c_str());
    if (orig_value) {
      orig_env_vars_[env_var] = orig_value;
    }
    setenv(env_var.c_str(), value.c_str(), 1);
  }
  orig_log_level_ = static_cast<int>(holoscan::log_level());
  nvidia::logger::GxfLogger& gxf_logger = nvidia::logger::GlobalGxfLogger::instance();
  orig_executor_log_level_ = static_cast<int>(gxf_logger.level());
}
EnvVarWrapper::EnvVarWrapper(std::string key, std::string value) : EnvVarWrapper({{key, value}}) {}

EnvVarWrapper::~EnvVarWrapper() {
  // Restore original environment variables
  for (const auto& [env_var, _] : env_var_settings_) {
    auto it = orig_env_vars_.find(env_var);
    if (it == orig_env_vars_.end()) {
      unsetenv(env_var.c_str());
    } else {
      setenv(env_var.c_str(), it->second.c_str(), 1);
    }
  }
  holoscan::set_log_level(static_cast<holoscan::LogLevel>(orig_log_level_));
  nvidia::logger::GxfLogger& gxf_logger = nvidia::logger::GlobalGxfLogger::instance();
  gxf_logger.level(static_cast<int>(orig_executor_log_level_));
}

File: python/tests/system/env_wrapper.py
"""
SPDX-FileCopyrightText: Copyright (c) 2023-2026 NVIDIA CORPORATION & AFFILIATES. All rights reserved.
SPDX-License-Identifier: Apache-2.0
"""  # noqa: E501

import os
import warnings
from contextlib import contextmanager

from holoscan.logger import log_level, set_log_level


@contextmanager
def env_var_context(env_var_settings=None):
    """Context manager for temporarily setting environment variables.

    Example usage:

    >>> env_var_settings = {
    ...     ("HOLOSCAN_CHECK_RECESSION_PERIOD_MS", "10"),
    ...     ("HOLOSCAN_MAX_DURATION_MS", "10000"),
    ...     ("HOLOSCAN_UCX_NETWORK_CONNECTION_TIMEOUT", "10000"),
    ...     ("HOLOSCAN_STOP_ON_DEADLOCK_TIMEOUT", "1000")
    ... }
    >>> with env_var_context(env_var_settings):
    ...     app = DistributedVideoReplayerApp()
    ...     app.run()
    """
    if env_var_settings is None:
        env_var_settings = {}

    # remember existing environment variables
    orig_env_vars = {env_var: os.environ.get(env_var) for env_var, _ in env_var_settings}
    orig_log_level = log_level()
    # Warning: The GXF logger level cannot be accessed or modified through Python APIs. If
    # 'HOLOSCAN_EXECUTOR_LOG_LEVEL' is modified here, we won't be able to restore it to its
    # original value when exiting this context.
    for env_var, value in env_var_settings:
        if env_var == "HOLOSCAN_EXECUTOR_LOG_LEVEL":
            # Skip setting GXF logger level as it's not accessible via Python APIs
            warnings.warn(
                "Cannot set HOLOSCAN_EXECUTOR_LOG_LEVEL through this context manager. "
                "The GXF logger level is not accessible through Python APIs, making it "
                "impossible to restore its original value when exiting this context.",
                stacklevel=2,
            )
            continue
        os.environ[env_var] = value
    try:
        yield
    finally:
        # restore the environment variable
        for env_var, _ in env_var_settings:
            orig_env = orig_env_vars.get(env_var)
            if orig_env is None:
                os.environ.pop(env_var, None)
            else:
                os.environ[env_var] = orig_env
        set_log_level(orig_log_level)
